File: analysis/sb-analyzer.jl
#!/usr/bin/env julia
using DataFrames, DataArrays, HypothesisTests, Distributions
#= TODO: CLI =#
#= read table from csv =#
df = readtable("data-format.csv")
# Assert data is complete and unique per Pid
#= for subdf in groupby(df, :Pid) =#
#=   println("pid check", size(subdf,1)) =#
#= end =#

# Fancy one-line data validation for participants ID
# TODO this could be done for other Task IDs aswell
println("Validating dataset...")
for field in [:Pid, :Tid]
  # Assert that all the Pid, Jid, groups have the same sizes
  chk = length(unique([size(subdf,1) for subdf in groupby(df, field)])) == 1
  println(field,": ", chk) 
end
burger_df = df[df[:navigation] .== "burger", :]
swipe_df = df[df[:navigation] .== "swipe", :]

burger = burger_df[:time_ms]
burger_mean, burger_std = mean_and_std(burger)
burger_isnorm_pval = pvalue(ExactOneSampleKSTest(burger, Normal(burger_mean, burger_std)))
burger_isnorm = burger_isnorm_pval < 0.05
println("Check Burger for normal distribution: $burger_isnorm [$burger_isnorm_pval]")
for swipe_task_df in groupby(swipe_df, :Tid)
  tid = swipe_task_df[1,:Tid]
  println("Swipe $tid against Burger")
  swipe_task = swipe_task_df[:time_ms]
  swipe_task_isnorm_pval = pvalue(ExactOneSampleKSTest(swipe_task, Normal(mean(swipe_task),std(swipe_task))))
  swipe_task_isnorm = swipe_task_isnorm_pval < 0.05
  println("Check Swipe Task $tid for normal distribution: $swipe_task_isnorm [$swipe_task_isnorm_pval]")
  if burger_isnorm && swipe_task_isnorm
    pval = pvalue(UnequalVarianceTTest(swipe_task, burger))
    println("  pvalue: $pval [Welch's T-Test]")
  else
    pval = pvalue(MannWhitneyUTest(swipe_task, burger))
    println("  pvalue: $pval [WMW U-Test]")
  end
end

File: analysis/sb-analyzer.py
#!/usr/bin/env python3
# -*- coding:utf8 -*-
""" Executable """
from __future__ import print_function
import argparse
import pandas as pd
from scipy import stats


def analyze_efficiency(df, verbose=0):
    nav_methods = df.groupby('navigation')
    burger_df, swipe_df = nav_methods.get_group('burger'), nav_methods.get_group('swipe')

    burger = burger_df["time_ms"]
    burger_shapiro = stats.shapiro(burger)
    burger_normality = burger_shapiro[1] < 0.05
    print("Checking Burger for normal distribution:", burger_normality, burger_shapiro)


    burger_tasks = [group["time_ms"] for _, group in burger_df.groupby("Tid")]
    # alternatives: repeated measures anova, friedmann test
    print("[BURGER] Kruskal-Wallis H-Test: ", stats.kruskal(*burger_tasks))
    print("[BURGER] Friedmann Chi Square: ", stats.friedmanchisquare(*burger_tasks))
    # FIXME do these tests work as expected? (they report low p values but the data is all the same)


    if verbose:
        print("Burger Navigation",burger.describe(), sep="\n")

    for tid, swipe_task_df in swipe_df.groupby("Tid"):
        swipe_task = swipe_task_df["time_ms"]
        if verbose:
            print("Swipe Navigation, Task %d" %tid, swipe_task.describe(), sep="\n")
        swipe_task_shapiro = stats.shapiro(swipe_task)
        swipe_task_normality = swipe_task_shapiro[1] < 0.05
        print ("Checking Task", tid, "for normal distribution:", swipe_task_normality, swipe_task_shapiro)

        result = stats.ttest_ind(swipe_task, burger, equal_var=False)\
            if (burger_normality and swipe_task_normality) else\
            stats.mannwhitneyu(swipe_task, burger)
        print(tid, ":", result, result[1] < 0.05)


def main():
    """ Main Function """
    parser = argparse.ArgumentParser()
    parser.add_argument("file", nargs='?',type=argparse.FileType('r'), default="data-format.csv")
    parser.add_argument("-v", "--verbose", action="count", default=0)
    args = parser.parse_args()
    df = pd.read_csv(args.file)

    # assert that we have the same number of samples for each Pid, Tid per navigation method, ...
    check = {str(field) : len(pd.unique([len(group) for _, group in df.groupby(field)])) == 1
             for field in ["Pid", ["navigation","Tid"]]}
    print("Dataset Validation:", check)
    analyze_efficiency(df, verbose=args.verbose)

    exit(0)


if __name__ == '__main__':
    main()
